File: icc.py
import numpy as np
import pandas as pd
import scipy as scp
import seaborn as sb
import sklearn as skl
import matplotlib as plot
import matplotlib.pyplot as plt
import scipy.stats as stats

import pingouin as pg

import argparse
import sys
import time
import multiprocessing as mp

def rearrange(sub, roi_name, conditions, inDir='/scratch/st-tv01-1/hcp/targets'):
    df = pd.DataFrame(columns = ['subject', 'run', 'value', 'edge', 'roi'])
    for cond in conditions:
        print(f'loading {sub} {cond}...')
        file = f'{inDir}/sub{sub}_{cond}_{roi_name}.csv'
        #print(file)
        matrix = pd.read_csv(file, sep=',', header=None)
        shape = matrix.values.flatten().shape[0]
        subject = np.full((shape,1), sub)
        run = np.full((shape,1), cond)
        values = np.transpose(np.array(matrix.values.flatten(), ndmin = 2))
        edge = np.transpose(np.array(range(0, shape), ndmin = 2))
        roi = np.full((shape,1), roi_name)
        df_to_add = pd.DataFrame(np.concatenate((subject, run, values, edge, roi), axis = 1), columns = ['subject', 'run', 'value', 'edge', 'roi'])
        df = pd.concat([df, df_to_add])
    return df

def icc(df):
    if df[0] % 50000 == 0:
        print(f'edge {df[0]}')
    return (df[0], pg.intraclass_corr(data = df[1], targets = 'subject', raters = 'run', ratings = 'value').ICC[4])
    

def main(logfile=None):
    if logfile is not None:
        f = open(logfile, 'w')
        sys.stdout = f
        
    roi = ['dlpfc', 'tpj', 'pre_sma']
    conditions = ['REST1', 'REST4', 'MOVIE2', 'MOVIE4']
    sub = pd.read_csv('/arc/project/st-tv01-1/hcp/targets/m2m4_sub_n111.csv', header = None).squeeze()
    # loop over each roi
    pool = mp.Pool(mp.cpu_count())
    for r in roi:
        for c in ['REST','MOVIE']:
            # setup starmap args
            args = []
            start = time.time()
            for s in sub:
                #print(f'loading {s}...')
                args.append((s, r, [x for x sin conditions if x.startswith(c)]))
            results = pool.starmap(rearrange, args)
            # concat subs to 1 df
            df = results.pop(0)
            while len(results) > 0:
                df = pd.concat((df, results.pop()))
            # run icc code and save
            minutes = round((time.time() - start) / 60,1)
            print(f'all {r} {c} data loaded in {minutes} min')
            start = time.time()
            df = list(df.groupby('edge'))
            icc = pool.map(icc_analysis, df)
            minutes = round((time.time() - start)/60,1)
            print(f'{r} {c} calculated in {minutes}')
            del(df)
            np.savetxt(f'/scratch/st-tv01-1/hcp/targets/icc_{r}_n{len(sub)}.txt', np.array(icc))
    pool.close()
    print(df.shape)
    print(df.head())
    

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("logfile", default=None)
    args = parser.parse_args()
    main(args.logfile)

File: icc.m
function icc(varargin)
repoDir = fileparts(mfilename('fullpath'));
try
    sub = load(fullfile(repoDir,'data','m2m4_sub_n111.csv'));
catch
    sub = [];
end
p = inputParser;
p.addParameter('fcDir','/scratch/st-tv01/hcp/targets');
p.addParameter('rois', {'tpj', 'dlpfc', 'pre_sma'});
p.addParameter('runs', {'REST1', 'REST4', 'MOVIE2','MOVIE4'});
p.addParameter('sub',sub);
p.addParameter('logfile','');
p.addParameter('cores',feature('numcores'));

p.parse(varargin{:});
inputs = p.Results;
fcDir = inputs.fcDir;
sub = inputs.sub;
rois = inputs.rois;
runs = inputs.runs;
cond = unique(cellfun(@(x) x(1:end-1),runs,'UniformOutput',0)); % get conditions
logfile = inputs.logfile;

if ~isempty(logfile)
    fid = fopen(logfile,'w');
else
    fid = 0;
end
% start parpool
util.start_parpool(inputs.cores);
% loop over rois
for i=1:numel(rois)
    % loop over condtions
    for j=1:numel(cond)
        % pick runs matching condition
        r = find(startsWith(runs,cond{j}));
        % preallocate arrays
        tmp = loadData(fcDir, sub(1), runs{end},rois{i});
        dim = size(tmp);
        edges = dim(1)*dim(2);
        run1 = zeros(numel(sub),edges,'single');
        run2 = zeros(numel(sub),edges,'single');
        % loop over and load subject data
        tic
        xprintf(fid,'%s %s\n', rois{i},cond{j});
        parfor s=1:numel(sub)
            tmp1 = loadData(fcDir, sub(s), runs{r(1)}, rois{i});
            tmp2 = loadData(fcDir, sub(s), runs{r(2)}, rois{i});
            run1(s,:) = reshape(tmp1,1,edges);
            run2(s,:) = reshape(tmp2,1,edges);
        end
        xprintf(fid,'\n\tsubjects loaded in %.1f minutes',toc/60);
        icc = zeros(edges,1);
        % loop over edges, calculate icc
        tic
        parfor e=1:size(run1,2)
            if ~mod(e, 100000)
                xprintf(fid,'\n\t\tedge %d',e);
            end
            icc(e) = util.icc21([run1(:,e) run2(:,e)]);
        end
        xprintf(fid,'\n\t%d edge iccs calculated in %.1f min',edges,toc/60);
        % save data
        file = fullfile(fcDir,sprintf('icc_%s_%s_n%d.txt',rois{i},cond{j}, numel(sub)));
        writematrix(icc,file);
        xprintf(fid,'%s written!',file);
    end
end
if fid
    fclose(fid);
end
end

function mat = loadData(matDir, sub, run, roi)
    mat = load(fullfile(matDir,sprintf('sub%d_%s_%s.csv', sub, run, roi)));
end

function xprintf(fid, varargin)
if fid
    fprintf(fid, varargin{:});
else
    fprintf(varargin{:});
end
end